### DISCUSSION TITLE
Sonoff Dongle Plus P and new FW

### DISCUSSION THREAD

**Discussion by Stianasv** [Koenkk/Z-Stack-firmware#514] [Q&A]
Is the latest firmware's from mid 2023 to the latest 2024.07.10 to large for the flash memory of the Sonoff Zigbee 3.0 Dongle Plus P (CC2652P) manufactured back in 2021?
Flashing older fw and the one I'm using today (2022.12.26) works great. But the newer one fails in Flash Programmer 2 and Zigstar. Not tried python yet..

**Comment by Nardol** [comment]
Have you tried with [cc2538-bsl](https://github.com/JelmerT/cc2538-bsl)?
It is probably not the kind of solution you expect but I remind I read some people having this kind of issue but were able to flash with cc2538-bsl.

EDIT: I should have read until the end, you have not tried yet but it could help you to try with this tool.

**Reply by Stianasv** [reply]
Not tried that yet...

But as I understand it looks like the P dongle only got 352kB programmable memory (https://www.ti.com/product/CC2652P). And the new FW is 494kB.. But again the FW I'm currently using (2022.12.26) is 400kB. 
I can't see anyone mention this in the latest fw feedback thread even it looks like others are using the same coordinator too. It's still listed as a tested/supported adapter but thats maybe not the case anymore with the latest firmware's.

**Reply by Hedda** [reply]
> Have you tried with [cc2538-bsl](https://github.com/JelmerT/cc2538-bsl)?

FYI, there are now also many forks of JelmerT's cc2538-bsl with improvements or new features/functions not yet merged to upstream:

* https://github.com/JelmerT/cc2538-bsl/issues/154

**Reply by Nardol** [reply]
Nice, even if we are now on a jungle of forks which make the choice very difficult...

**Comment by tomsdk** [comment]
I have successfully updated the Sonoff Dongle P using TI flash tool. I did have to convert hex-file to bin-file first, though.
To be exact, I updated two units, both to 20240710 and both CC2652P.

**Comment by Hedda** [comment]
Just make sure you download and flash the correct image file as different adapters use different images depending on pins and chip used:

https://github.com/Koenkk/Z-Stack-firmware/blob/master/coordinator/Z-Stack_3.x.0/bin/README.md

For the SONOFF Zigbee 3.0 USB Dongle Plus by ITead you are doing to need to use the `CC1352P2_CC2652P_launchpad_*.zip` file:

https://github.com/Koenkk/Z-Stack-firmware/releases

https://github.com/Koenkk/Z-Stack-firmware/releases/download/Z-Stack_3.x.0_coordinator_20240710/CC1352P2_CC2652P_launchpad_coordinator_20240710.zip

https://github.com/Koenkk/Z-Stack-firmware/raw/master/coordinator/Z-Stack_3.x.0/bin/CC1352P2_CC2652P_launchpad_coordinator_20240710.zip

PS: I have also seen many make the mistake of downloading the HTML page from GitHub instead of downloaded the raw ZIP file.

**Comment by Wickedy123** [comment]
I managed to flash 2024-07-10 using the Texas Programmer tool without any issues - what error are you getting?

Things I had to do:
- Download the drivers as I have a new PC. Prior to this, it appeared as a bad device in Device Manager and was not auto-detected by the TI programmer. Specifically the Universal Windows Driver from https://www.silabs.com/developers/usb-to-uart-bridge-vcp-drivers?tab=downloads
- I ticked the 'Erase' box as well in TI - not sure I always did this in the past. 
- Convert HEX to BIN

**Reply by Stianasv** [reply]
I got "Unknown record type: 3" error in the TI flash program trying to flash the hex file.. Without checking to much I'm pretty sure its because it's not big enough to fit the file..

Don't know what you used to convert it with. But I used hex2bin.py in python to convert it. It created a bin file 352kB large.. Flashed that to the sonoff dongle with TI flash programmer and it worked great.

Thanks for pointing me in the right direction guys:)

**Comment by Dillton** [comment]
Hi.
I found somewhere on internet the simplest method how to update firmware on these dongles is ZigStar TI CC2652P/P7 FW Flasher add-on for HomeAssistant. I use this method since last year and it's perfect.
![image](https://github.com/user-attachments/assets/393c4642-b79f-4734-b257-a69594c4b5ef)

So if you can use it just use it and you won't need to disconnect and disassemble the dongle every time when there is a firmware update and do the update on another PC. 

Below you can see my notes which steps you need to take and how to configure the add-on correctly for Sonoff dongle P.
![image](https://github.com/user-attachments/assets/e2adc02d-e70c-46c1-9dba-022adef36a23)

Hope this makes life easier for some pople.

**Reply by cvroque** [reply]
I used to flash the dongle on my computer, but this worked flawlessly, thanks for the tip!

**Reply by bloy911** [reply]
Чувак ты гений! Спасибо тебе!

**Reply by sdetweil** [reply]
@Dillton  you should mention that if you use HA in docker, you don't get this tool and cannot install it.. 
another dead end

**Reply by Simonz82** [reply]
> ZigStar TI CC2652P/P7 FW Flasher

> I hope this makes life easier for some people.

it is not clear to me, I have a SONOFF ZigBee 3.0 USB Dongle Plus, can I use this tool to update the firmware, do I still have the 20210708 version ?
because in the documentation it says:

"For USB attached coordinators:
In the add-on configuration enter some text into the Network Device entry form to make the form happy
Select the correct serial device from the list of devices detect, Toggle the switch to enable USB Flash (Coordinator must be in bootloader mode before addon is started, hold BSL button while plugging into USB) Not applicable to Sonoff.
Enable required firmware toggle or enter the url for the Z-Stack Firmware - be sure to use the raw GitHub link example: [copy](https://github.com/Koenkk/Z-Stack-firmware/raw/7398d834eb3a790876c280293c4181da96cc7114/coordinator/Z-Stack_3.x.0/bin/CC1352P2_CC2652P_launchpad_coordinator_20221226.zip)
Start the add-on.
Monitor the flash procedure by going to the addon log page, and refreshing periodically."

**Reply by jeffgoh** [reply]
FWIW https://dongle.sonoff.tech/sonoff-dongle-flasher/ worked great for me. My "Sonoff Zigbee 3.0 USB Dongle Plus" aka ZBDongle-P as detected on the site. I updated from 20210708 to 20250321 and works perfectly with both z2m 2.5.1 and z2m 2.9.1.

Z2M feels quite a bit faster than before, but not sure if that's from the new firmware or just from rebooting.
